packages/server/src/internals/middlewares.ts

```diff
@@ -0,0 +1,31 @@
+import { ProcedureType } from '../router';
+import { TRPCError } from '../TRPCError';
+
+export const middlewareMarker = Symbol('middlewareMarker');
+interface MiddlewareResultBase {
+  /**
+   * All middlewares should pass through their `next()`'s output.
+   * Requiring this marker makes sure that can't be forgotten at compile-time.
+   */
+  readonly marker: typeof middlewareMarker;
+}
+
+interface MiddlewareOKResult extends MiddlewareResultBase {
+  ok: true;
+  data: unknown;
+  // this could be extended with `input`/`rawInput` later
```

Comment:
`input` would be on the `opts` of `MiddlewareFunction`, not here. Tbh I was convinced by your argument and don't think we even need this comment.